### ISSUE TITLE
amsmath: alternative spacing of columns in align

### ISSUE THREAD

**Issue by nbeisert** [latex3/latex2e#224]
## Brief outline of the bug

The `align` environment of `amsmath` is a math block that produces columns which are split evenly on the available horizontal space. The `aligned` environment is a math element that produces columns which are separated by a fixed horizontal space. There seems to be no environment that is a stand-alone math block with the column placement behaviour of `aligned`, and there seems to be no option or combination to achieve this behaviour without undesired side-effects. I would like to suggest an enhancement to `amsmath` to fill this gap.

## Further Information

- `aligned` is not a suitable work-around because it has no numbering for individual lines and cannot be broken across pages.
- for narrow columns the separating space is very large on a wide line; a fixed separation for `align` can make the appearance more uniform.
- same layout for `equation`, `aligned` and `align`; makes them somewhat interchangable.

Potential extensions:

- option in `amsmath` to globally change the behaviour for `align` (and similar environments)
- introduce an optional maximum separation for `align` columns
- new environment(s) for aligned columns with fixed separation

## Minimal example showing the bug

```latex
\RequirePackage{latexbug}
\documentclass{article}

\RequirePackage{amsmath}

\renewcommand{\minalignsep}{2em}

\makeatletter
\expandafter\def\expandafter\calc@shift@align\expandafter
  {\expandafter\hook@align@sep\calc@shift@align}
\def\aligncolumnsorg{\def\hook@align@sep{}}
\aligncolumnsorg
\def\aligncolumnsalt{
 \def\hook@align@sep{\ifnum\xatlevel@=\@ne
  \dimen@\displaywidth
  \advance\dimen@-\totwidth@
  \@tempcntb\maxfields@
  \divide\@tempcntb\tw@
  \@tempcnta\@tempcntb
  \advance\@tempcntb\m@ne
  \global\eqnshift@\dimen@
  \global\alignsep@\minalignsep\relax
  \global\advance\eqnshift@-\@tempcntb\alignsep@
  \global\divide\eqnshift@\tw@
  \ifdim\eqnshift@<\z@
   \global\eqnshift@\z@
  \fi
 \fi}
}
\makeatother

\begin{document}

equation:
\begin{equation}
a^2+b^2=c^2 \hspace{\minalignsep} x^2+y^2=z^2
\end{equation}

aligned:
\begin{equation}
|\begin{aligned}
a^2+b^2&=c^2 & x^2+y^2&=z^2
\end{aligned}|
%unrelated: is the space after the last column intended?
%or rather: can it be eliminated somehow?
\end{equation}

align (original columns): 
\begin{align}
a^2+b^2&=c^2 & x^2+y^2&=z^2
\end{align}

align (alternative columns): 
\aligncolumnsalt
\begin{align}
a^2+b^2&=c^2 & x^2+y^2&=z^2
\end{align}

\end{document}
```
This is merely a hack of the existing `amsmath` code to illustrate a potential extension and what settings would have to tweaked to achieve the desired behaviour. Depending on whether and which extension will be considered, I could try to contribute more suitable code.

## log/pdf files

[amsmathcol.pdf](https://github.com/latex3/latex2e/files/3878977/amsmathcol.pdf)
[amsmathcol.log](https://github.com/latex3/latex2e/files/3878979/amsmathcol.log)

**Comment by RuixiZhang42** [comment]
Not very necessary IMHO. Have you tried `alignat`?

```latex
\RequirePackage{latexbug}
\documentclass{article}

\usepackage{amsmath}

\begin{document}

\begin{alignat}{2}
a^2+b^2&=c^2 & \qquad x^2+y^2&=z^2
% \qquad = horizontal skip of 2em
\end{alignat}

\end{document}
```

**Comment by nbeisert** [comment]
Thanks for your reply and suggestion. Unfortunately that would require _a lot_ of manual interaction by setting the spacing manually for each line and each column (recall that entries have different widths, so many `\qquad`s will have no effect). It also mixes layout with content which would be nice to avoid. I'd argue the alternative behaviour has as much justification as the established one, and either one will be preferrable in some situations, hence it would be great to have the option of choice.

**Comment by hvoss49** [comment]
You can use the (deprecated) environment `xalignat`:

```
\documentclass{article}
\usepackage{amsmath}
\begin{document}

\begin{xalignat}{2}
a^2+b^2&=c^2 &  x^2+y^2&=z^2
% \qquad = horizontal skip of 2em
\end{xalignat}

\end{document}
```

**Comment by RuixiZhang42** [comment]
The problem with the extra space in `aligned` was reported by me a while back: https://github.com/latex3/latex2e/issues/160

```latex
\RequirePackage{latexbug}
\documentclass{article}

\usepackage{amsmath}

\begin{document}

equation:
\begin{equation}
a^2+b^2=c^2 \hspace{\minalignsep} x^2+y^2=z^2
\end{equation}

aligned broken:
\begin{equation}
|\begin{aligned}
a^2+b^2&=c^2 & x^2+y^2&=z^2
\end{aligned}|
\end{equation}

aligned temporary fix:
\begin{equation}
|\begin{aligned}
a^2+b^2&=c^2 & x^2+y^2&=z^2\\ % <- add a line break here
\end{aligned}|
\end{equation}

\end{document}
```

Judging from https://github.com/latex3/latex2e/issues/160 this is not going to be fixed in the foreseeable future `:(`

**Comment by nbeisert** [comment]
@hvoss49 : thanks, but this gives the same output as `align`.

Internally, the alignment envirnments are serviced by `\start@align` with 3 parameters. The first parameter is `\xatlevel@` which can apparently be:
- `0`: no space between columns
- `1`: columns split evenly among avaliable space
- `2`: full length variant

the second parameter specifies the starred variant. the third parameter specifies the number of columns or `-1` for automatic determination (essentially `{align}` is `{xalignat}{-1}`).
One implementation could be to change the behaviour of mode `0` to _fixed space between columns_: This would mimic the behaviour of `aligned` (serviced by `\start@aligned`) which adds `\minalignsep` between the columns. In order not to break existing documents, one would have to choose a different variable name for use in `\start@align` and set it to `0pt` initially. Then one could use `{alignat}{-1}` (or better define a new environment to avoid having to specify `-1`).

Another implementation would be a mode 3 which does just that.

**Comment by RuixiZhang42** [comment]
@nbeisert Mode 3 would be an excellent idea!

```latex
\documentclass{article}

\usepackage{amsmath}
\renewcommand{\minalignsep}{2em}
\usepackage{etoolbox}

\makeatletter
\patchcmd{\measure@}{%
            \global\divide\alignsep@\@tempcntb
}{%
            \global\divide\alignsep@\@tempcntb
        \or
            \global\alignsep@\minalignsep\relax
            \@tempcntb\maxfields@
            \divide\@tempcntb\tw@
            \@tempcnta\@tempcntb
            \advance\@tempcntb\m@ne
            \if@fleqn
                \global\eqnshift@\@mathmargin
            \else
                \global\advance\@tempcnta\@ne
                \global\eqnshift@\dimen@
                \global\advance\eqnshift@-\@tempcntb\alignsep@
                \global\divide\eqnshift@\tw@
            \fi
}{%
  \typeout{Allow new align structures}%
}{%
  \typeout{Failed to patch \string\measure@}%
}

\newenvironment{newalign}{%
  \start@align\thr@@\st@rredfalse\m@ne
}{%
  \endalign
}
\newenvironment{newalign*}{%
  \start@align\thr@@\st@rredtrue\m@ne
}{%
  \endalign
}
\makeatother

\begin{document}

equation:
\begin{equation}
a^2+b^2=c^2 \hspace{\minalignsep} x^2+y^2=z^2
\end{equation}

aligned:
\begin{equation}
|\begin{aligned}
a^2+b^2&=c^2 & x^2+y^2&=z^2\\ % <- eliminate extra space with a line break
\end{aligned}|
\end{equation}

align (original columns): 
\begin{align}
a^2+b^2&=c^2 & x^2+y^2&=z^2
\end{align}

new align (original columns): 
\begin{newalign}
a^2+b^2&=c^2 & x^2+y^2&=z^2
\end{newalign}

\end{document}
```

**Comment by nbeisert** [comment]
@RuixiZhang42 Yes, that'd be a marvellous enhancement! As far as I can see, the code is consistent with my hack, and that worked well for a couple of documents I've been using it on. What would be a suitable name? `falign` (fixed)? `sepalign`? `calign` (compact/center)? Cannot come up with a convincing one...

**Comment by nbeisert** [comment]
Here's another solution using a maximum separation `\maxalignsep`. This one has some benefit on its own right, i.e. it could be considered in parallel, as one can use it in combination with `\minalignsep` to compress when there is insufficient space. 

```
\documentclass{article}

\usepackage{amsmath}
\renewcommand{\minalignsep}{1em}
\usepackage{etoolbox}

\makeatletter

\patchcmd{\measure@}{%
        \fi
        \ifdim\eqnshift@<\z@
}{%
        \fi
        \ifdim\maxalignsep<\z@\else
            \ifdim\alignsep@>\maxalignsep\relax
                \global\alignsep@\maxalignsep\relax
                \if@fleqn\else
                    \global\eqnshift@\displaywidth
                    \global\advance\eqnshift@-\totwidth@
                    \global\advance\eqnshift@-\@tempcntb\alignsep@
                    \global\divide\eqnshift@\tw@
                \fi
            \fi
        \fi
        \ifdim\eqnshift@<\z@
}{%
  \typeout{Introduce maximum column separation}%
}{%
  \typeout{Failed to patch \string\measure@}%
}

\newcommand{\maxalignsep}{-1pt}

\makeatother

\begin{document}

align:
\begin{align}
a^2+b^2&=c^2 & a^2+b^2&=c^2
\end{align}
\begin{align}
a^2+b^2&=c^2 & a^2+b^2&=c^2 & a^2+b^2&=c^2
\end{align}
\begin{align}
a^2+b^2&=c^2 & a^2+b^2&=c^2 & a^2+b^2&=c^2 & a^2+b^2&=c^2
\end{align}
\begin{align}
a^2+b^2&=c^2 & a^2+b^2&=c^2 & a^2+b^2&=c^2 & a^2+b^2&=c^2 & 
a^2+b^2&=c^2
\end{align}
\begin{align}
a^2+b^2&=c^2 & a^2+b^2&=c^2 & a^2+b^2&=c^2 & a^2+b^2&=c^2 & 
a^2+b^2&=c^2 & a^2+b^2&=c^2
\end{align}

\renewcommand{\maxalignsep}{2em}

align with maxalignsep: 
\begin{align}
a^2+b^2&=c^2 & a^2+b^2&=c^2
\end{align}
\begin{align}
a^2+b^2&=c^2 & a^2+b^2&=c^2 & a^2+b^2&=c^2
\end{align}
\begin{align}
a^2+b^2&=c^2 & a^2+b^2&=c^2 & a^2+b^2&=c^2 & a^2+b^2&=c^2
\end{align}
\begin{align}
a^2+b^2&=c^2 & a^2+b^2&=c^2 & a^2+b^2&=c^2 & a^2+b^2&=c^2 & 
a^2+b^2&=c^2
\end{align}
\begin{align}
a^2+b^2&=c^2 & a^2+b^2&=c^2 & a^2+b^2&=c^2 & a^2+b^2&=c^2 & 
a^2+b^2&=c^2 & a^2+b^2&=c^2
\end{align}

\end{document}
```

**Comment by nbeisert** [comment]
Please disregard the above proposal concerning `\maxalignsep`. It has undesired behaviour for `flalign` and a better solution is given in the separate issue #230.
